src/kernels/ipywidgets-message-coordination/cdnWidgetScriptSourceProvider.ts

```diff
@@ -0,0 +1,294 @@
+import { Telemetry } from '../../platform/common/constants';
+import { IConfigurationService, IHttpClient, WidgetCDNs } from '../../platform/common/types';
+import { createDeferred, Deferred } from '../../platform/common/utils/async';
+import { Common, DataScience } from '../../platform/common/utils/localize';
+import { noop } from '../../platform/common/utils/misc';
+import { traceError, traceInfo, traceVerbose } from '../../platform/logging';
+import { ConsoleForegroundColors } from '../../platform/logging/types';
+import { sendTelemetryEvent } from '../../telemetry';
+import { IWidgetScriptSourceProvider, WidgetScriptSource } from './types';
+
+// Source borrowed from https://github.com/jupyter-widgets/ipywidgets/blob/54941b7a4b54036d089652d91b39f937bde6b6cd/packages/html-manager/src/libembed-amd.ts#L33
+const unpgkUrl = 'https://unpkg.com/';
+const jsdelivrUrl = 'https://cdn.jsdelivr.net/npm/';
+
+export const GlobalStateKeyToTrackIfUserConfiguredCDNAtLeastOnce = 'IPYWidgetCDNConfigured';
+export const GlobalStateKeyToNeverWarnAboutScriptsNotFoundOnCDN = 'IPYWidgetNotFoundOnCDN';
+export const GlobalStateKeyToNeverWarnAboutNoNetworkAccess = 'IPYWidgetNoNetWorkAccess';
+
+function moduleNameToCDNUrl(cdn: string, moduleName: string, moduleVersion: string) {
+    let packageName = moduleName;
+    let fileName = 'index'; // default filename
+    // if a '/' is present, like 'foo/bar', packageName is changed to 'foo', and path to 'bar'
+    // We first find the first '/'
+    let index = moduleName.indexOf('/');
+    if (index !== -1 && moduleName[0] === '@') {
+        // if we have a namespace, it's a different story
+        // @foo/bar/baz should translate to @foo/bar and baz
+        // so we find the 2nd '/'
+        index = moduleName.indexOf('/', index + 1);
+    }
+    if (index !== -1) {
+        fileName = moduleName.substr(index + 1);
+        packageName = moduleName.substr(0, index);
+    }
+    if (cdn === jsdelivrUrl) {
+        // Js Delivr doesn't support ^ in the version. It needs an exact version
+        if (moduleVersion.startsWith('^')) {
+            moduleVersion = moduleVersion.slice(1);
+        }
+        // Js Delivr also needs the .js file on the end.
+        if (!fileName.endsWith('.js')) {
+            fileName = fileName.concat('.js');
+        }
+    }
+    return `${cdn}${packageName}@${moduleVersion}/dist/${fileName}`;
+}
+
+function getCDNPrefix(cdn?: WidgetCDNs): string | undefined {
+    switch (cdn) {
+        case 'unpkg.com':
+            return unpgkUrl;
+        case 'jsdelivr.com':
+            return jsdelivrUrl;
+        default:
+            break;
+    }
+}
+/**
+ * Widget scripts are found in CDN.
+ * Given an widget module name & version, this will attempt to find the Url on a CDN.
+ * We'll need to stick to the order of preference prescribed by the user.
+ */
+export class CDNWidgetScriptSourceProvider implements IWidgetScriptSourceProvider {
```

Comment:
Merged the web and node files into one, we don't need to download the cdn files into the local file system anymore.
If we do that then things will not work, e.g. ipyvolume and a few other widgets don't work with that approach.

If user doesn't have access to the internet from the webview, then we'll not use CDN, this way we're consistent. I.e. we always use CDN when there's network connection, else fallback.
If here's no network access we're now displaying a warning message.